### DISCUSSION TITLE
How do I move my old recipes into new version?

### DISCUSSION THREAD

**Discussion by ChrisGroenewald** [Brewtarget/brewtarget#754] [Q&A] [answered]
I have recipe's/equipment set up in 2.3.0.

I tried v3 recently but got the crash on new recipe, so I waited for a fix which landed today. Thanks!

But, I'm confused about how I'm supposed to get my recipes into v3.0.9. When I start it up I'm back to a blank slate with only the Bt recipe's visible. I tried to export a recipe to xml from 2.3.0 and import to 3.0.9, but that fails. I tried to backup my DB and restore it in 3... but get a warning that everything (I'm assuming this includes new fermentables, etc that would be great to have) will be wiped. I don't want that either.

So I'm not sure how I reconcile my old recipes into the new version. Bit confusing, or just not possible?

On Windows, FWIW...

Cheers.

**Comment by matty0ung** [comment]
Sorry to hear you're having problems.  In theory, what should happen when you install a newer version is that it automatically upgrades the old 2.3.0 database for you.  What I'm wondering is whether 3.0.9 is looking for the database in a different place than 2.3.0.  (If the software can't find a database file, it creates a new one for you with the default recipes.)

Hopefully, all you need to do is find the old brewtarget 2.3.0 database file and copy it over the default-generated 3.0.9 database file.  You're looking for a file called `database.sqlite`.  The first places to look are inside subfolders of the hidden `AppData` folder inside your user folder.  (Eg on my machine I look in `C:\Users\Matt\AppData\`. ) Check in all three subfolders, though I think you should usually find what you're looking for in `Local`.  I'm hoping you'll find two folders, one called something like `brewtarget_2.3.0` and another called `brewtarget`.  Make sure Brewtarget is not running.  Then, rename `brewtarget/database.sqlite` to `brewtarget/database.sqlite.bak` and copy `brewtarget_2.3.0/database.sqlite` into the `brewtarget` folder.

This is partly from memory, so give us a shout if what you find is totally different!

**Reply by ChrisGroenewald** [reply]
Thanks, it was: No, yes, and yes, and then no.

I looked and there was only one in 'brewtarget'. I had to hunt around to find the other in the, fairly weird ..../AppData/Local/VirtualStore/Program Files (x86)/Brewtarget-2.3.0/data/ ... But Ok, did find it.

I noted it's smaller (I only have a couple recipes) than the 3.x db. Suspecting my intuition would be right: I backed up the 3.x db, copied the 2.3 over, and yes... there are my recipes. But I note many other things disappear (obviously) like fermentables, yeasts, "Bt" recipes .... Which isn't great.

I've restored the 3.x one now and will probably, I guess, get a sqlite3 client and try do this the hard way.

2c: A better migration path would be nice.

Thanks.

**Comment by matty0ung** [comment, marked as answer]
Glad you found the DB files!  This means there should be an easier way to get your old recipes.

The "export to XML (aka BeerXML)" functionality in 2.3.0 had a few bugs, which should be addressed in newer versions of the software.  You _should_ be able to get your old recipes from 2.3.0 as follows:
 * Start up the 3.0.9 version of the software using the 2.3.0 database (as you did before)
 * Export your recipes to one or more BeerXML files
 * Exit the software, restore the 3.x database and restart so you're running the new software with the new DB
 * Import the recipes from the BeerXML files

Let us know how you get on.

**Reply by ChrisGroenewald** [reply]
Hi, been a bit slammed, but had a moment, +thanks to your prompt, to sit down and actually do this: Worked out great. Somehow all my "BT" recipes got duplicated "(1)" ... I'm not sure what I did to do that, but I could just remove them all again.

Anywho, I've been able to successfully migrate so I'll call that a win. Thanks for the great software! Looking forward to your improvements below ... got a batch "crafted" in BT in the fermenter right now, so Cheers again!

**Comment by matty0ung** [comment]
> 2c: A better migration path would be nice.

Agreed.  I'll try to remember to have a look about adding new data when upgrading from old databases.  There is a bit of stuff already that pulls in new data from BeerXML.  I want to revamp that when we get to BeerJSON and provide everyone a lot more fermentables, yeasts, hops, default recipes etc.

**Comment by matty0ung** [comment]
The various bugs above should be fixed in current versions of the software, so archiving this discussion.
